src/Silverback.Integration/Messaging/Broker/Consumer.cs

```diff
@@ -1,59 +1,109 @@
-        where TBroker : class, IBroker
-        where TEndpoint : class, IEndpoint
-    {
-        protected Consumer(IBroker broker, IEndpoint endpoint,
-            ILogger<Consumer> logger, MessageLogger messageLogger)
-            : base(broker, endpoint, logger, messageLogger)
-        {
-        }
-
-        protected new TBroker Broker => (TBroker)base.Broker;
-
-        protected new TEndpoint Endpoint => (TEndpoint)base.Endpoint;
-    }
-}
\ No newline at end of file
+﻿// Copyright (c) 2018-2019 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+using Silverback.Messaging.Messages;
+
+namespace Silverback.Messaging.Broker
+{
+    public abstract class Consumer : EndpointConnectedObject, IConsumer
+    {
+        private const string DiagnosticListenerName = "Silverback.Messaging.Broker.Consumer";
+        private const string ActivityName = "Silverback.Messaging.Broker.Consumer.ConsumeMessage";
+        private const string CorrelationIdHeaderKey = "CorrelationId"; // TODO: Config.
+        private const string CorrelationBaggageHeaderKey = "CorrelationBaggage"; // TODO: Config.
+
+        private readonly ILogger<Consumer> _logger;
+        private readonly MessageLogger _messageLogger;
+        private readonly DiagnosticListener _diagnosticListener;
+
+        protected Consumer(IBroker broker, IEndpoint endpoint, ILogger<Consumer> logger, MessageLogger messageLogger, DiagnosticListener diagnosticListener)
+           : base(broker, endpoint)
+        {
+            _logger = logger;
+            _messageLogger = messageLogger;
+            _diagnosticListener = diagnosticListener;
+        }
+
+        public event EventHandler<MessageReceivedEventArgs> Received;
+
+        public void Acknowledge(IOffset offset)
+        {
+            Acknowledge(new[] { offset });
+        }
+
+        public abstract void Acknowledge(IEnumerable<IOffset> offsets);
+
+        protected void HandleMessage(byte[] message, IEnumerable<MessageHeader> headers, IOffset offset)
+        {
+            Activity activity = StartActivity(headers);
+
+            if (Received == null)
+                throw new InvalidOperationException("A message was received but no handler is configured, please attach to the Received event.");
+
+            var deserializedMessage = Endpoint.Serializer.Deserialize(message);
+
+            _messageLogger.LogTrace(_logger, "Message received.", deserializedMessage, Endpoint);
+
+            Received.Invoke(this, new MessageReceivedEventArgs(deserializedMessage, headers, offset));
+
+            StopActivity(activity);
+        }
+
+        private Activity StartActivity(IEnumerable<MessageHeader> messageHeaders)
+        {
+            string correlationId =  messageHeaders.GetFromHeaders(CorrelationIdHeaderKey);

```

Comment:
Here actually the magic happens.
If there is a correlationIn the header, its used as Parent.
If not then just a new Activity without a parentId is created.
From here on, the ActivityId flows through the system.